CRM/Utils/Check/Component/DedupeRules.php

```diff
@@ -0,0 +1,105 @@
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+class CRM_Utils_Check_Component_DedupeRules extends CRM_Utils_Check_Component {
+
+  /**
+   * Get dedupe rules, grouped by contact type, for a specific usage
+   *
+   * @param string $used (Supervised, Unsupervised, General)
+   * @return string[]
+   */
+  private static function getContactTypesForRule($used) {
+    $dedupeRules = \Civi\Api4\DedupeRuleGroup::get()
+      ->addSelect('contact_type')
+      ->setGroupBy([
+        'contact_type',
+      ])
+      ->addWhere('used', '=', $used)
+      ->execute();
+
+    $types = [];
+    foreach ($dedupeRules as $rule) {
+      if (!in_array($rule, $types)) {
+        $types[] = $rule['contact_type'];
+      }
+    }
+    return $types;
+  }
+
+  /**
+   * Returns an array of missing expected contact types
+   *
+   * @param string[] $rules
+   * @return array
+   */
+  private static function getMissingRules($rules) {
+    $types = ['Organization', 'Household', 'Individual'];
+    return array_diff($types, $rules);
+  }
+
+  /**
+   * @return CRM_Utils_Check_Message[]
+   */
+  public static function checkSupervisedRulesExist() {
+    $supervisedRules = self::getContactTypesForRule('Supervised');
+    $missingRules = self::getMissingRules($supervisedRules);
+    if (empty($missingRules)) {
+      return [];
+    }
+
+    $message = new CRM_Utils_Check_Message(
+      __FUNCTION__,
+      ts('For CiviCRM to function correctly you must have a supervised dedupe rule configured for each contact type. You are missing a supervised rule for: %1', [1 => implode(', ', $missingRules)]),
+      ts('Supervised dedupe rules missing'),
+      \Psr\Log\LogLevel::WARNING,
+      'fa-server'
+    );
+    $message->addAction(
+      ts('Configure dedupe rules'),
+      FALSE,
+      'href',
+      ['path' => 'civicrm/contact/deduperules', 'query' => ['reset' => 1]]
+    );
+    return [$message];
+  }
+
+  /**
+   * @return CRM_Utils_Check_Message[]
+   */
+  public static function checkUnsupervisedRulesExist() {
+    $unsupervisedRules = self::getContactTypesForRule('Unsupervised');
+    $missingRules = self::getMissingRules($unsupervisedRules);
+    if (empty($missingRules)) {
+      return [];
+    }
+
+    $message = new CRM_Utils_Check_Message(
+      __FUNCTION__,
+      ts('For CiviCRM to function correctly you must have an unsupervised dedupe rule configured for each contact type. You are missing an unsupervised rule for: %1', [1 => implode(', ', $missingRules)]),
+      ts('Unsupervised dedupe rules missing'),
+      \Psr\Log\LogLevel::WARNING,
+      'fa-server'
+    );
+    $message->addAction(
+      ts('Configure dedupe rules'),
+      FALSE,
+      'href',
+      ['path' => 'civicrm/contact/deduperules', 'query' => ['reset' => 1]]
+    );
+    return [$message];
```

Comment:
I understand the need for tidyness, but I prefer to avoid unnecessary code duplication whenever possible. What about this:

```suggestion
  public static function checkDedupeRulesExist() {
    $messages = [];
    
    $ruleTypes = ['Supervised' => ts('Supervised'), 'Unsupervised' => ts('Unsupervised')];
    foreach ($ruleTypes as $ruleType => $ruleLabel) {
      $rules = self::getContactTypesForRule($ruleType);
      $missingRules = self::getMissingRules($rules);
      if ($missingRules) {
        $messages[] = (new CRM_Utils_Check_Message(
          __FUNCTION__ . $ruleType,
          ts('For CiviCRM to function correctly you must have at least 2 dedupe rules configured for each contact type. You are missing a %1 rule for: %2', [1 => $ruleLabel, 2 => implode(', ', $missingRules)]),
          ts('%1 dedupe rules missing', [1 => $ruleLabel]),
          \Psr\Log\LogLevel::WARNING,
          'fa-server'
        ))->addAction(
          ts('Configure dedupe rules'),
          FALSE,
          'href',
          ['path' => 'civicrm/contact/deduperules', 'query' => ['reset' => 1]]
        );
      }
    }
    return $messages;
  }
```

Reply:
Yep this looks good. I just had to make a tiny tweak to your suggestion for it to run correctly, which I've just pushed.